File: examples/python/task_queue_example.py
#!/usr/bin/env python3
"""
FastCollection v1.0.0 - Task Queue Example (Python)

Implements a persistent task queue with priorities and dead letter queue.

Usage: python task_queue_example.py

Copyright © 2025-2030, Ashutosh Sinha (ajsinha@gmail.com)
Patent Pending
"""

from fastcollection import FastQueue
import json
import time
import random
from dataclasses import dataclass, asdict
from typing import Optional


@dataclass
class Task:
    id: str
    task_type: str
    payload: dict
    priority: int = 5  # 0 = highest priority
    created_at: float = None
    retry_count: int = 0
    max_retries: int = 3
    
    def __post_init__(self):
        if self.created_at is None:
            self.created_at = time.time()
    
    def to_json(self) -> bytes:
        return json.dumps(asdict(self)).encode()
    
    @classmethod
    def from_json(cls, data: bytes) -> 'Task':
        return cls(**json.loads(data.decode()))
    
    def should_retry(self) -> bool:
        return self.retry_count < self.max_retries


class TaskQueue:
    def __init__(self, base_path: str, task_ttl: int = 3600):
        self.queue = FastQueue(f"{base_path}/tasks.fc")
        self.dlq = FastQueue(f"{base_path}/dlq.fc")
        self.task_ttl = task_ttl
    
    def submit(self, task: Task):
        """Submit a task to the queue"""
        if task.priority == 0:
            # High priority - add to front
            self.queue.offer_first(task.to_json(), ttl=self.task_ttl)
        else:
            self.queue.offer(task.to_json(), ttl=self.task_ttl)
        print(f"Submitted: {task.id} ({task.task_type}) priority={task.priority}")
    
    def poll(self) -> Optional[Task]:
        """Get next task from queue"""
        data = self.queue.poll()
        if data is None:
            return None
        return Task.from_json(data)
    
    def peek(self) -> Optional[Task]:
        """Peek at next task without removing"""
        data = self.queue.peek()
        if data is None:
            return None
        return Task.from_json(data)
    
    def requeue(self, task: Task):
        """Requeue a failed task or move to DLQ"""
        task.retry_count += 1
        
        if task.should_retry():
            print(f"  Requeuing: {task.id} (attempt {task.retry_count}/{task.max_retries})")
            self.queue.offer(task.to_json(), ttl=self.task_ttl)
        else:
            print(f"  Moving to DLQ: {task.id} (max retries exceeded)")
            self.dlq.offer(task.to_json(), ttl=-1)  # Keep forever in DLQ
    
    def size(self) -> int:
        return self.queue.size()
    
    def dlq_size(self) -> int:
        return self.dlq.size()
    
    def close(self):
        self.queue.close()
        self.dlq.close()
    
    def __enter__(self):
        return self
    
    def __exit__(self, *args):
        self.close()


def process_task(task: Task) -> bool:
    """Simulate task processing with random failures"""
    print(f"Processing: {task.id} ({task.task_type})")
    time.sleep(0.1)  # Simulate work
    
    # 30% chance of failure
    if random.random() < 0.3:
        print(f"  FAILED!")
        return False
    
    print(f"  SUCCESS!")
    return True


def main():
    print("FastCollection v1.0.0 - Task Queue Example (Python)")
    print("=" * 55)
    print()
    
    with TaskQueue("/tmp/taskqueue_py", task_ttl=3600) as tq:
        # Submit various tasks
        print("Submitting tasks...\n")
        
        tasks = [
            Task("t1", "EMAIL", {"to": "user@example.com", "subject": "Welcome"}),
            Task("t2", "REPORT", {"type": "daily", "format": "pdf"}, priority=3),
            Task("t3", "ALERT", {"message": "Critical system alert!"}, priority=0),
            Task("t4", "BACKUP", {"database": "users", "destination": "s3"}, priority=5),
            Task("t5", "NOTIFY", {"user_id": "123", "message": "New message"}, priority=1),
        ]
        
        for task in tasks:
            tq.submit(task)
        
        print(f"\nQueue size: {tq.size()}")
        
        next_task = tq.peek()
        if next_task:
            print(f"Next task: {next_task.id} ({next_task.task_type})")
        
        # Process tasks
        print("\n--- Processing Tasks ---\n")
        
        while True:
            task = tq.poll()
            if task is None:
                break
            
            if not process_task(task):
                tq.requeue(task)
        
        print("\n--- Summary ---")
        print(f"Queue size: {tq.size()}")
        print(f"Dead letter queue size: {tq.dlq_size()}")


if __name__ == "__main__":
    main()

File: examples/java/TaskQueueExample.java
/**
 * FastCollection v1.0.0 - Task Queue Example
 * 
 * Implements a persistent task queue with priorities using FastCollectionQueue.
 * 
 * Copyright © 2025-2030, Ashutosh Sinha (ajsinha@gmail.com)
 * Patent Pending
 */
package examples;

import com.kuber.fastcollection.*;
import java.io.Serializable;
import java.util.*;
import java.util.concurrent.*;

public class TaskQueueExample {
    
    public static class Task implements Serializable {
        private static final long serialVersionUID = 1L;
        
        public String id;
        public String type;
        public String payload;
        public int priority;  // 0 = highest
        public long createdAt;
        public int retryCount;
        public int maxRetries;
        
        public Task(String id, String type, String payload, int priority) {
            this.id = id;
            this.type = type;
            this.payload = payload;
            this.priority = priority;
            this.createdAt = System.currentTimeMillis();
            this.retryCount = 0;
            this.maxRetries = 3;
        }
        
        @Override
        public String toString() {
            return String.format("Task{id=%s, type=%s, priority=%d, retries=%d/%d}",
                id, type, priority, retryCount, maxRetries);
        }
    }
    
    public static class TaskQueue {
        private final FastCollectionQueue<Task> queue;
        private final FastCollectionQueue<Task> deadLetterQueue;
        private final int taskTTL;
        
        public TaskQueue(String basePath, int taskTTLSeconds) {
            this.queue = new FastCollectionQueue<>(basePath + "/tasks.fc", 64 * 1024 * 1024, true);
            this.deadLetterQueue = new FastCollectionQueue<>(basePath + "/dlq.fc", 16 * 1024 * 1024, true);
            this.taskTTL = taskTTLSeconds;
        }
        
        public void submit(Task task) {
            if (task.priority == 0) {
                // High priority - add to front
                queue.offerFirst(task, taskTTL);
            } else {
                queue.offer(task, taskTTL);
            }
            System.out.println("Submitted: " + task);
        }
        
        public Task poll() {
            return queue.poll();
        }
        
        public Task peek() {
            return queue.peek();
        }
        
        public void requeue(Task task) {
            task.retryCount++;
            if (task.retryCount >= task.maxRetries) {
                System.out.println("Moving to DLQ: " + task);
                deadLetterQueue.offer(task, -1);  // Keep forever in DLQ
            } else {
                System.out.println("Requeuing: " + task);
                queue.offer(task, taskTTL);
            }
        }
        
        public int size() {
            return queue.size();
        }
        
        public int deadLetterSize() {
            return deadLetterQueue.size();
        }
        
        public void close() {
            queue.close();
            deadLetterQueue.close();
        }
    }
    
    public static void main(String[] args) throws InterruptedException {
        System.out.println("FastCollection v1.0.0 - Task Queue Example");
        System.out.println("==========================================\n");
        
        TaskQueue taskQueue = new TaskQueue("/tmp/taskqueue", 3600);
        
        try {
            // Submit various tasks
            System.out.println("Submitting tasks...\n");
            
            taskQueue.submit(new Task("t1", "EMAIL", "Send welcome email to user@example.com", 2));
            taskQueue.submit(new Task("t2", "REPORT", "Generate daily sales report", 3));
            taskQueue.submit(new Task("t3", "ALERT", "Critical system alert!", 0));  // High priority
            taskQueue.submit(new Task("t4", "BACKUP", "Backup user database", 5));
            taskQueue.submit(new Task("t5", "NOTIFY", "Push notification to mobile", 1));
            
            System.out.println("\nQueue size: " + taskQueue.size());
            System.out.println("Next task: " + taskQueue.peek());
            
            // Process tasks
            System.out.println("\n--- Processing Tasks ---\n");
            
            Random random = new Random();
            Task task;
            
            while ((task = taskQueue.poll()) != null) {
                System.out.println("Processing: " + task);
                
                // Simulate random failures (30% chance)
                if (random.nextInt(10) < 3) {
                    System.out.println("  FAILED! Requeuing...");
                    taskQueue.requeue(task);
                } else {
                    System.out.println("  SUCCESS!");
                }
                
                Thread.sleep(100);  // Simulate work
            }
            
            System.out.println("\n--- Summary ---");
            System.out.println("Queue size: " + taskQueue.size());
            System.out.println("Dead letter queue size: " + taskQueue.deadLetterSize());
            
        } finally {
            taskQueue.close();
        }
    }
}
